pkg/machine/compression/decompress.go

```diff
@@ -1,279 +1,114 @@
-	// }
-	// if isZip && runtime.GOOS == "windows" {
-	// 	return decompressZip(prefix, localPath.GetPath(), uncompressedFileWriter)
-	// }
+const (
+	zipExt            = ".zip"
+	progressBarPrefix = "Extracting compressed file"
+	macOs             = "darwin"
+)
 
-	//  Unfortunately GZ is not sparse capable.  Lets handle it differently
-	// if compressionType == archive.Gzip && runtime.GOOS == "darwin" {
-	// 	return decompressGzWithSparse(prefix, localPath, uncompressedPath)
-	// }
-	// return decompressEverythingElse(prefix, localPath.GetPath(), uncompressedFileWriter)
+type decompressor interface {
+	srcFilePath() string
+	reader() (io.Reader, error)
+	copy(w *os.File, r io.Reader) error
+	close()
 }
 
-// Will error out if file without .Xz already exists
-// Maybe extracting then renaming is a good idea here..
-// depends on Xz: not pre-installed on mac, so it becomes a brew dependency
-func decompressXZ(prefix string, src string, output io.WriteCloser) error {
-	var read io.Reader
-	var cmd *exec.Cmd
-
-	stat, err := os.Stat(src)
-	if err != nil {
-		return err
-	}
-	file, err := os.Open(src)
-	if err != nil {
-		return err
-	}
-	defer file.Close()
-
-	p, bar := utils.ProgressBar(prefix, stat.Size(), prefix+": done")
-	proxyReader := bar.ProxyReader(file)
-	defer func() {
-		if err := proxyReader.Close(); err != nil {
-			logrus.Error(err)
-		}
-	}()
-
-	// Prefer Xz utils for fastest performance, fallback to go xi2 impl
-	if _, err := exec.LookPath("xz"); err == nil {
-		cmd = exec.Command("xz", "-d", "-c")
-		cmd.Stdin = proxyReader
-		read, err = cmd.StdoutPipe()
-		if err != nil {
-			return err
-		}
-		cmd.Stderr = os.Stderr
-	} else {
-		// This XZ implementation is reliant on buffering. It is also 3x+ slower than XZ utils.
-		// Consider replacing with a faster implementation (e.g. xi2) if podman machine is
-		// updated with a larger image for the distribution base.
-		buf := bufio.NewReader(proxyReader)
-		read, err = xz.NewReader(buf)
-		if err != nil {
-			return err
-		}
-	}
-
-	done := make(chan bool)
-	go func() {
-		if _, err := io.Copy(output, read); err != nil {
-			logrus.Error(err)
-		}
-		output.Close()
-		done <- true
-	}()
-
-	if cmd != nil {
-		err := cmd.Start()
-		if err != nil {
-			return err
-		}
-		p.Wait()
-		return cmd.Wait()
+func newDecompressor(compressedFilePath string, compressedFileContent []byte) decompressor {
+	compressionType := archive.DetectCompression(compressedFileContent)
+	os := runtime.GOOS
+	hasZipSuffix := strings.HasSuffix(compressedFilePath, zipExt)
+
+	switch {
+	case compressionType == archive.Xz:
+		return newXzDecompressor(compressedFilePath)
+	case compressionType == archive.Uncompressed && hasZipSuffix:
+		return newZipDecompressor(compressedFilePath)
+	case compressionType == archive.Uncompressed:
+		return newUncompressedDecompressor(compressedFilePath)
+	case compressionType == archive.Gzip && os == macOs:
```

Comment:
This is the current behavior. I haven't changed that. I am assuming that the gzip implementation in `c/images/pkg/compression` on doesn't support sparse files on macos. And we use `crc/pkg/os` implementation instead.

Reply:
AFAICS `CopySparse` consumes a plain `io.Reader`, I can’t see an obvious reason why it couldn’t consume (the faster one used by) the `c/image/pkg/compression` one.

Either way, maybe not this PR.

Reply:
For the time being I have added the following comment:
```
	// macOS gzipped VM images are sparse. As a result a
	// special decompressor is required: it uses crc os.CopySparse
	// instead of io.Copy and std lib gzip instead of klauspost/pgzip
	// (even if it's slower).
```

Reply:
I’m not too much of a fan of the platform dependency, but, \*shrug\*, preexisting.

---

> I suspect that this is a problem with CopySparse. Anyway, not something I would fix in this PR.

Fair enough, but please at least file an issue if you are not already working on this; silent data corruption seems serious.

It’s quite possible that it would be fixed by using #21763 instead.

Reply:
Right, let me try with #21763 because I suspect the problem was crc CopySparse indeed.

Reply:
Using the `SparseWriter` introduced in #21763 fixed the problem. So I have been able to update this PR to use the gzip decompressor from `c/image/pkg/compression` (that uses `klauspost/pgzip`) and it's faster 🚀 . Although `zstd` is used now rather than `gzip` (through `klauspost/compress/zstd`) but that's fast too.